packages/std/src/math/decimal.rs

```diff
@@ -118,16 +118,33 @@ impl Decimal {
 
     /// Returns the ratio (numerator / denominator) as a Decimal
     pub fn from_ratio(numerator: impl Into<Uint128>, denominator: impl Into<Uint128>) -> Self {
+        match Decimal::checked_from_ratio(numerator, denominator) {
+            Ok(value) => value,
+            Err(CheckedFromRatioError::DivideByZero) => {
+                panic!("Denominator must not be zero")
+            }
+            Err(CheckedFromRatioError::MulOverflow) => panic!("Multiplication overflow"),
+        }
+    }
+
+    /// Returns the ratio (numerator / denominator) as a Decimal
+    pub fn checked_from_ratio(
+        numerator: impl Into<Uint128>,
+        denominator: impl Into<Uint128>,
+    ) -> Result<Self, CheckedFromRatioError> {
         let numerator: Uint128 = numerator.into();
         let denominator: Uint128 = denominator.into();
         if denominator.is_zero() {
-            panic!("Denominator must not be zero");
+            return Err(CheckedFromRatioError::DivideByZero);
         }
 
-        Decimal(
-            // numerator * DECIMAL_FRACTIONAL / denominator
-            numerator.multiply_ratio(Self::DECIMAL_FRACTIONAL, denominator),
-        )
+        match numerator.checked_multiply_ratio(Self::DECIMAL_FRACTIONAL, denominator) {
+            Ok(ratio) => {
+                // numerator * DECIMAL_FRACTIONAL / denominator
+                Ok(Decimal(ratio))
+            }
+            Err(_) => Err(CheckedFromRatioError::MulOverflow),
```

Comment:
```suggestion
            Err(CheckedMultiplyRatioError::DivideByZero) => Err(CheckedFromRatioError::DivideByZero),
            Err(CheckedMultiplyRatioError::MulOverflow) => Err(CheckedFromRatioError::MulOverflow),
```

if we write it like this, we don't need the zero check above

Reply:
Good point. So having whole `CheckedFromRatioError` as an enum doesn't make sense either, since it will be only one case (overflow) which is inherited from other enum anyway.